modules/seqexec/server/src/main/scala/seqexec/server/StepsView.scala

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2016-2019 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package seqexec.server
+
+import cats.implicits._
+import cats.data.NonEmptyList
+import seqexec.model.ActionType
+import seqexec.model.Step
+import seqexec.model.StandardStep
+import seqexec.model.NodAndShuffleStep
+import seqexec.model.NodAndShuffleStatus
+import seqexec.model.dhs.ImageFileId
+import seqexec.model.enum.ActionStatus
+import seqexec.model.enum.Resource
+import seqexec.model.enum.Instrument
+import seqexec.model.enum.Instrument._
+import seqexec.model.enum.SystemName
+import seqexec.model.StepState
+import seqexec.engine
+import seqexec.engine.Action.ActionState
+import seqexec.engine.Action
+import seqexec.engine.ParallelActions
+import seqexec.server.gmos.Gmos
+
+sealed trait StepsView[F[_]] {
+  /**
+   * This method creates a view of the step for the client
+   * The Step returned maybe a StandardStep of be specialized e.g. for N&S
+   */
+  def stepView(
+    stepg: SequenceGen.StepGen[F],
+    step: engine.Step[F],
+    altCfgStatus: List[(Resource, ActionStatus)]
+  ): Step
+}
+
+object StepsView {
+  private def kindToResource(kind: ActionType): List[Resource] = kind match {
+    case ActionType.Configure(r) => List(r)
+    case _                       => Nil
+  }
+
+  def splitAfter[A](l: List[A])(p: A => Boolean): (List[A], List[A]) =
+    l.splitAt(l.indexWhere(p) + 1)
+
+  private[server] def separateActions[F[_]](ls: NonEmptyList[Action[F]]): (List[Action[F]], List[Action[F]]) =
+    ls.toList.partition(_.state.runState match {
+        case ActionState.Completed(_) => false
+        case ActionState.Failed(_)    => false
+        case _                        => true
+      }
+    )
+
+  private[server] def configStatus[F[_]](executions: List[ParallelActions[F]]): List[(Resource, ActionStatus)] = {
+    // Remove undefined actions
+    val ex = executions.filter { !separateActions(_)._2.exists(_.kind === ActionType.Undefined) }
+    // Split where at least one is running
+    val (current, pending) = splitAfter(ex)(separateActions(_)._1.nonEmpty)
+
+    // Calculate the state up to the current
+    val configStatus = current.foldLeft(Map.empty[Resource, ActionStatus]) {
+      case (s, e) =>
+        val (a, r) = separateActions(e).bimap(
+            _.flatMap(a => kindToResource(a.kind).tupleRight(ActionStatus.Running)).toMap,
+            _.flatMap(r => kindToResource(r.kind).tupleRight(ActionStatus.Completed)).toMap)
+        s ++ a ++ r
+    }
+
+    // Find out systems in the future
+    val presentSystems = configStatus.keys.toList
+    // Calculate status of pending items
+    val systemsPending = pending.map {
+      s => separateActions(s).bimap(_.map(_.kind).flatMap(kindToResource), _.map(_.kind).flatMap(kindToResource))
```

Comment:
Worthwhile abstracting this operation out? It occurs two more times in `pendingConfigStatus`.